dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -3921,41 +3921,40 @@ Spdp::process_lease_expirations(const DCPS::MonotonicTimePoint& now)
 {
   ACE_GUARD (ACE_Thread_Mutex, g, lock_);
 
-  for (TimeQueue::iterator pos = lease_expirations_.begin(), limit = lease_expirations_.end();
-       pos != limit && pos->first <= now;) {
+  // Process one at a time to keep the thread responsive.
+  TimeQueue::iterator pos = lease_expirations_.begin();
+  if (pos != lease_expirations_.end() && pos->first <= now) {
     DiscoveredParticipantIter part = participants_.find(pos->second);
     // Pre-emptively erase so purge_discovered_participant will not modify lease_expirations_.
     lease_expirations_.erase(pos++);
 
-    if (part == participants_.end()) {
-      continue;
-    }
-
-    if (DCPS::DCPS_debug_level) {
-      ACE_DEBUG((LM_WARNING,
-                 ACE_TEXT("(%P|%t) Spdp::process_lease_expirations() - ")
-                 ACE_TEXT("participant %C exceeded lease duration, removing\n"),
-                 DCPS::LogGuid(part->first).c_str()));
-    }
+    if (part != participants_.end()) {
+      if (DCPS::DCPS_debug_level) {
+        ACE_DEBUG((LM_DEBUG, "(%P|%t) DEBUG: Spdp::process_lease_expirations: "
+                   "participant %C exceeded lease duration, removing\n",
+                   DCPS::LogGuid(part->first).c_str()));
+      }
 
 #if OPENDDS_CONFIG_SECURITY
-    DCPS::WeakRcHandle<ICE::Endpoint> sedp_endpoint = sedp_->get_ice_endpoint();
-    if (sedp_endpoint) {
-      stop_ice(sedp_endpoint, part->first, part->second.pdata_.participantProxy.availableBuiltinEndpoints,
-               part->second.pdata_.participantProxy.availableExtendedBuiltinEndpoints);
-    }
-    DCPS::WeakRcHandle<ICE::Endpoint> spdp_endpoint = tport_->get_ice_endpoint();
-    if (spdp_endpoint) {
-      ice_agent_->stop_ice(spdp_endpoint, guid_, part->first);
-    }
-    purge_handshake_deadlines(part);
+      DCPS::WeakRcHandle<ICE::Endpoint> sedp_endpoint = sedp_->get_ice_endpoint();
+      if (sedp_endpoint) {
+        stop_ice(sedp_endpoint, part->first, part->second.pdata_.participantProxy.availableBuiltinEndpoints,
+                 part->second.pdata_.participantProxy.availableExtendedBuiltinEndpoints);
+      }
+      DCPS::WeakRcHandle<ICE::Endpoint> spdp_endpoint = tport_->get_ice_endpoint();
+      if (spdp_endpoint) {
+        ice_agent_->stop_ice(spdp_endpoint, guid_, part->first);
+      }
+      purge_handshake_deadlines(part);
 #endif
-    purge_discovered_participant(part);
-    participants_.erase(part);
+      purge_discovered_participant(part);
+      participants_.erase(part);
+    }
   }
 
   if (!lease_expirations_.empty()) {
-    tport_->lease_expiration_task_->schedule(lease_expirations_.begin()->first - now);
+    const TimeDuration delay = std::max(lease_expirations_.begin()->first - now, TimeDuration::from_msec(1));
```

Comment:
Wouldn't there be cases where a shorter min delay constant would be desired?  If there are many expirations, and they can be processed quickly, requiring 1ms can keep more data in memory than needed.

Also how about resampling the clock instead of using the passed-in `now`?  When determining the time for the next event, the passed-in time may not be that helpful.